packages/controllers/src/permissions/Caveat.test.ts

```diff
@@ -0,0 +1,60 @@
+import * as errors from './errors';
+import {
+  CaveatSpecificationConstraint,
+  CaveatSpecificationMap,
+  decorateWithCaveats,
+  PermissionConstraint,
+} from '.';
+
+// This function is mostly tested through the PermissionController tests,
+// we just add a couple here for completeness.
+describe('decorateWithCaveats', () => {
+  it('decorates a method with caveat', async () => {
+    const methodImplementation = () => [1, 2, 3];
+
+    const caveatSpecifications = {
+      reverse: {
+        decorator: (method: any, _caveat: any) => async () => {
+          return (await method()).reverse();
+        },
+      },
+    } as unknown as CaveatSpecificationMap<CaveatSpecificationConstraint>;
```

Comment:
Ah, I got rid of almost all casts here: a374667

The problem we're having with the `NonEmptyArray` type is that it is actually a type matching only non-empty _tuples_. I don't think the TypeScript array type encodes anything about how many members it contains.

For now, I got around this by casting just the caveats array as `[CaveatConstraint]`, which seems OK. Consumers are very unlikely to ever work with permission objects in this manner.

A better solution would be to assign the permission `as const`, however then `decorateWithCaveats` complains because the `PermissionConstraint` type (which it expects as a param) is not deeply `Readonly` at the moment. Bah, humbug. We can do something about that later.

Reply:
Ah, a better immediate solution presented itself: Just do `const permission: PermissionConstraint = { ... }`. 4f226a5

Reply:
Aha, tuples, of course. That makes sense. The solution you landed on seems good to me!